### PR TITLE
4492 Change :profile dynamic segment to :survey dynamic segment 

### PR DESCRIPTION
**Pull request by TylerMatteo**
### Summary
Moves the logic in the old `/profile` route over to `/survey` and updates all of the code to return all data (for all profiles) for the given survey (acs or decennial) and geoid.

#### Tasks/Bug Numbers
 - Fixes [AB#4218](https://dev.azure.com/NYCPlanning/cc280b0d-40a0-4689-b852-2e6247f1af50/_workitems/edit/4218)

### CODE REVIEW

File: routes/survey.js [hunk 1]
```diff
@@ -235,9 +239,10 @@ function sortRowByVariable(rowA, rowB) {
  * @param{string} profile - The profile type (TODO: this parameter's domain should be [`decennial`, `acs`]
  * @returns{function}
  */
-function getQueryBuilder(profile) {
-  // if (profile === 'decennial') return decennialQuery;
-  return profileQuery;
+function getQueryBuilder(survey) {
+  console.log({survey})
```
Comment (godfreyyeung) [inline routes/survey.js:243]: delete this

File: routes/survey.js [hunk 2]
```diff
@@ -102,32 +104,34 @@ router.get('/:id/', async (req, res) => {
...
+  const queryBuilder = getQueryBuilder(survey);
+  console.log(queryBuilder(geoids))
+  
   // get data from postgres
-  const [rawProfileData, rawCompareData, rawPreviousProfileData, rawPreviousCompareData] = await Promise.all([
-    db.query(queryBuilder(profileName, geoids)),
-    db.query(queryBuilder(profileName, [compareTo])),
-    db.query(queryBuilder(profileName, geoids, /* is previous */ true)),
-    db.query(queryBuilder(profileName, [compareTo], /* is previous */ true)),
+  const [rawProfileData, rawCompareProfileData, rawPreviousProfileData, rawPreviousCompareProfileData] = await Promise.all([
```
Comment (godfreyyeung) [inline routes/survey.js:120]: why add `Profile` token into these variable names? `Profile` here, like in `rawProfileData`, refers to the data for the selected geography. `rawCompareData` refers to data for the comparison geometry.
Actually I think it might be better to rename these as `rawSelectionData`, `rawCompareData`, `rawPreviousSelectionData`, `rawPreviousCompareData`

File: routes/survey.js [hunk 3]
```diff
@@ -46,18 +46,18 @@ function prefixObj(row, prefix) {
...
 
-router.get('/:id/', async (req, res) => {
+router.get('/:survey/:id/', async (req, res) => {
   const { app } = req;
-  let { id: _id } = req.params;
+  const { id: _id, survey } = req.params;
   const { compareTo = '0' } = req.query;
 
-  let [ idPrefix, selectionId ] = _id.split('_');
+  let [idPrefix, selectionId] = _id.split('_'); // eslint-disable-line
```
Comment (godfreyyeung) [inline routes/survey.js:54]: wait, why prefer using `// eslint-disable-line` here?

File: bin/www [hunk 4]
```diff
@@ -16,7 +16,7 @@ const app = require('../app');
  * Get port from environment and store in Express.
  */
 
-const port = (process.env.NODE_ENV === 'production') ? process.env.PORT : 3000;
+const port = (process.env.NODE_ENV === 'production') ? process.env.PORT : 3001;
```
Comment (godfreyyeung) [inline bin/www:19]: Is this necessary for this PR? Why this change?

File: query-helpers/profile.js [hunk 5]
```diff
@@ -21,7 +21,7 @@ function isAggregate(ids) {
   return ids.length > 1;
 }
 
-const profileSQL = (profile, ids, isPrevious = false) => `
+const profileSQL = (ids, isPrevious = false) => `
```
Comment (godfreyyeung) [inline query-helpers/profile.js:24]: Isn't `query-helpers/acs.js` replacing this file? Can't this file be deleted?

Comment (godfreyyeung) [reply query-helpers/profile.js:24]: Nit: Prefer that the file rename command is used in git to track the file renaming.

File: query-helpers/acs.js [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+      --- category ---
+      REGEXP_REPLACE(
+        LOWER(category),
+        '[^A-Za-z0-9]', '_', 'g'
+      ) AS category,
+      --- domain ---
+      REGEXP_REPLACE(
+        LOWER(domain),
+        '[^A-Za-z0-9]', '_', 'g'
+      ) AS domain,
```
Comment (godfreyyeung) [inline query-helpers/acs.js:145]: Did this not work before? I noticed you changed this from `profile` to `domain`, I'm guessing to match the `domain` column in the ACS tables. Did you change column schema that or was that just something we overlooked?

**Review by godfreyyeung — CHANGES_REQUESTED** [review]

File: query-helpers/acs.js (hunk 6 above)
Comment (godfreyyeung) [reply query-helpers/acs.js:145]: Just looked at the migration files, looks like it was `domain` this whole time. Do you know if this `domain` property is used in our API code anywhere?

File: routes/survey.js (hunk 3 above)
Comment (TylerMatteo) [reply routes/survey.js:54]: I actually added that because eslint was complaining about `idPrefix` being a `let` because it doesn't get reassigned, but I'll definitley put the spaces back

File: query-helpers/acs.js (hunk 6 above)
Comment (TylerMatteo) [reply query-helpers/acs.js:145]: I don't see it being used anywhere in the front end, I just changed this to keep things consistent with the schema.

File: routes/survey.js (hunk 3 above)
Comment (godfreyyeung) [reply routes/survey.js:54]: ohhh gotcha

File: query-helpers/acs.js (hunk 6 above)
Comment (godfreyyeung) [reply query-helpers/acs.js:145]: Gotcha

File: routes/survey.js [hunk 7]
```diff
@@ -102,32 +104,34 @@ router.get('/:id/', async (req, res) => {
...
  */
-async function getProfileData(profileName, geoids, compareTo, db) {
+// async function getProfileData(profileName, geoids, compareTo, db) {
+async function getSurveyData(survey, geoids, compareTo, db) {
   const isAggregate = geoids.length > 1;
 
-  const queryBuilder = getQueryBuilder(profileName);
-
+  const queryBuilder = getQueryBuilder(survey);
+  console.log(queryBuilder(geoids))
```
Comment (godfreyyeung) [inline routes/survey.js:117]: delete this guy

File: routes/survey.js [hunk 8]
```diff
@@ -169,10 +173,10 @@ function join(profileName, current, compare, previous, previousCompare) {
...
-    const rowConfig = find(specialCalculationConfigs[profileName], ['variable', row.variable]);
-    const compareRow = compare.find(compare => compare.id === row.id);
-    const previousRow = previous.find(previous => previous.id === row.id);
-    const previousCompareRow = previousCompare.find(previousCompare => previousCompare.id === row.id);
+    const rowConfig = find(specialCalculationConfigs[survey], ['variable', row.variable]);
+    const compareRow = compare.find(c => c.id === row.id);
+    const previousRow = previous.find(p => p.id === row.id);
+    const previousCompareRow = previousCompare.find(p => p.id === row.id);
 
     const difference = doDifferenceCalculations(row, compareRow);
```
Comment (godfreyyeung) [inline routes/survey.js:181]: Wait you added a new `isDecennial` parameter to these three functions, but didn't pass in an argument to that parameter....

**Review by godfreyyeung — CHANGES_REQUESTED** [review]

File: routes/survey.js [hunk 9]
```diff
@@ -77,16 +79,16 @@ router.get('/:id/', async (req, res) => {
         if (selection && selection.length > 0) {
           // TODO: remove "profile" argument, and corresponding parameter in upstream functions
           // TODO: What happens if there is more than one selected geography result?
-          profileObj = await getProfileData("demographic", selection[0].geoids, compareTo, app.db);
+          profileObj = await getSurveyData(survey, selection[0].geoids, compareTo, app.db);
```
[the remark this answers is not in the document]
Comment (TylerMatteo) [reply routes/survey.js:82]: Are you referring to `survey` or the `getSurveyData`? The former is just a string

File: routes/survey.js (hunk 8 above)
Comment (godfreyyeung) [reply routes/survey.js]: This was why the difference calculations weren't showing up

**Comment by godfreyyeung** [comment]
Core changes in this PR have been merged to develop through https://github.com/NYCPlanning/labs-factfinder-api/pull/132.
This PR eventually needs to be rebased onto develop again

File: routes/survey.js [hunk 10]
```diff
@@ -95,26 +95,29 @@ router.get('/:survey/:id/', async (req, res) => {
...
-// async function getProfileData(profileName, geoids, compareTo, db) {
 async function getSurveyData(survey, geoids, compareTo, db) {
   const isAggregate = geoids.length > 1;
 
   const queryBuilder = getQueryBuilder(survey);
+<<<<<<< HEAD
   
+=======
+
+>>>>>>> 52e491a947161264778cfb0e9c943e1ec4232a65
```
Comment (godfreyyeung) [inline routes/survey.js]: eek

File: routes/survey.js [hunk 11]
```diff
@@ -171,23 +174,30 @@ function join(survey, current, compare, previous, previousCompare) {
...
 
+<<<<<<< HEAD
     const difference = doDifferenceCalculations(row, compareRow, isDecennial);
     const previousDifference = doDifferenceCalculations(previousRow, previousCompareRow, isDecennial);
     const changeOverTime = doChangeCalculations(row, previousRow, rowConfig, isDecennial);
+=======
+    const difference = doDifferenceCalculations(row, compareRow, survey === 'decennial');
+    const previousDifference = doDifferenceCalculations(previousRow, previousCompareRow, survey === 'decennial');
+    const changeOverTime = doChangeCalculations(row, previousRow, rowConfig, survey === 'decennial');
+>>>>>>> 52e491a947161264778cfb0e9c943e1ec4232a65
```
Comment (godfreyyeung) [inline routes/survey.js]: looks like we want the second one?

File: routes/survey.js [hunk 12]
```diff
@@ -77,75 +79,79 @@ router.get('/:id/', async (req, res) => {
...
+async function getSurveyData(survey, geoids, compareTo, db) {
   const isAggregate = geoids.length > 1;
 
-  const queryBuilder = getQueryBuilder(profileName);
+  const queryBuilder = getQueryBuilder(survey);
+<<<<<<< HEAD
+  
+=======
 
+>>>>>>> 52e491a947161264778cfb0e9c943e1ec4232a65
```
Comment (godfreyyeung) [inline routes/survey.js:120]: eek

File: routes/survey.js [hunk 13]
```diff
@@ -168,23 +174,30 @@ function join(profileName, current, compare, previous, previousCompare) {
...
+
+<<<<<<< HEAD
+    const difference = doDifferenceCalculations(row, compareRow, isDecennial);
+    const previousDifference = doDifferenceCalculations(previousRow, previousCompareRow, isDecennial);
+    const changeOverTime = doChangeCalculations(row, previousRow, rowConfig, isDecennial);
+=======
+    const difference = doDifferenceCalculations(row, compareRow, survey === 'decennial');
+    const previousDifference = doDifferenceCalculations(previousRow, previousCompareRow, survey === 'decennial');
+    const changeOverTime = doChangeCalculations(row, previousRow, rowConfig, survey === 'decennial');
+>>>>>>> 52e491a947161264778cfb0e9c943e1ec4232a65
```
Comment (godfreyyeung) [inline routes/survey.js:192]: we want the second one?

File: routes/selection.js [hunk 14]
```diff
@@ -105,8 +105,13 @@ router.post('/', async (req, res) => {
   body.geoids.sort();
 
   const {
+<<<<<<< HEAD
```
Comment (godfreyyeung) [inline routes/selection.js:108]: should accept these changes from HEAD

File: query-helpers/acs.js [hunk 15]
```diff
@@ -138,10 +138,15 @@ const acsProfileSQL = (ids, isPrevious = false) => `
...
+      --- survey ---
+      'acs' AS survey
     FROM enriched_profile
-    GROUP BY variable, variablename, base, category, profile
+    GROUP BY variable, variablename, base, category, domain
```
Comment (godfreyyeung) [inline query-helpers/acs.js:149]: According to the old SQL, here `survey` should be included in this `GROUP BY` arguments list. Do you know why we should only include `domain` instead here?

File: routes/survey.js (hunk 2 above)
Comment (godfreyyeung) [reply routes/survey.js:109]: I think you accidentally re-included `Profile` here

File: routes/survey.js [hunk 16]
```diff
@@ -45,104 +45,99 @@ function prefixObj(row, prefix) {
...
 
   // Instantiate DataProcessors to process query results
-  const profileData = new DataProcessor(rawProfileData, profileName, isAggregate).process();
-  const compareData = new DataProcessor(rawCompareData, profileName, /* isAggregate */ false).process();
-  const previousProfileData = new DataProcessor(rawPreviousProfileData, profileName, isAggregate, /* isPrevious */ true).process();
-  const previousCompareData = new DataProcessor(rawPreviousCompareData, profileName, /* isAggregate */ false, /* isPrevious */ true).process();
+  const surveyData = new DataProcessor(rawProfileData, survey, isAggregate).process();
+  const compareSurveyData = new DataProcessor(rawCompareProfileData, survey, /* isAggregate */ false).process();
+  const previousSurveyData = new DataProcessor(rawPreviousProfileData, survey, isAggregate, /* isPrevious */ true).process();
+  const previousCompareSurveyData = new DataProcessor(rawPreviousCompareProfileData, survey, false, /* isPrevious */ true).process();
```
Comment (godfreyyeung) [inline routes/survey.js:120]: Should get rid off all the `profile` symbols here

File: query-helpers/acs.js (hunk 15 above)
Comment (TylerMatteo) [reply query-helpers/acs.js:149]: Why do you think `survey` should be included? In this case `profile` is analogous to `domain`, not `survey`. Also, in this SQL `survey` is hard coded as `acs` so I'm not sure what including it in a group by would accomplish (begs the question if we even need to add that `survey` field at all).

Comment (godfreyyeung) [reply query-helpers/acs.js:149]: It looked like to me that profile is analogous to survey here, not domain. Since profile used to hold the value 'acs', and now survey holds the value 'acs'.
